### PR TITLE
Improve types for (un)compiled template results

### PR DESCRIPTION
**Pull request by rictic**
Adds two new types: `UncompiledTemplateResult` and `MaybeCompiledTemplateResult`. Currently `UncompiledTemplateResult` is the same as `TemplateResult`, and `MaybeCompiledTemplateResult` is the union of the compiled and uncompiled types.

We'd like for code to be able to handle the `html` and `svg` functions returning a MaybeCompiledTemplateResult. Changing that type will be a major breaking change though, and we don't want to do a major version so quickly, but this lays the ground work, and provides future-compatible types that people can use when they want to be explicit about which type they mean, because while TemplateResult will change, MaybeCompiled, Uncompiled, and Compiled will all have the same meanings in the next major version.

### CODE REVIEW

File: packages/lit-html/src/lit-html.ts [hunk 1]
```diff
@@ -466,13 +466,38 @@ const COMMENT_PART = 7;
...
+ * and type of template (html or svg).
+ *
+ * `TemplateResult` objects do not create any DOM on their own. To create or
+ * update DOM you need to render the `TemplateResult`. See
+ * [Rendering](https://lit.dev/docs/components/rendering) for more information.
+ *
+ * In Lit 4, this type will be an alias of
+ * MaybeCompiledTemplateResult, so that code will get type errors if it assumes
+ * that Lit templates are not compiled. When deliberately working with only
+ * one, use either CompiledTemplateResult or UncompiledTemplateResult
```
Comment (AndrewJakubowicz) [inline packages/lit-html/src/lit-html.ts:495]: Nit: Could make `CompiledTemplateResult` and `UncompiledTemplateResult` use `@linkcode` so they link to the respective docs.

File: packages/lit-html/src/lit-html.ts [hunk 2]
```diff
@@ -466,13 +466,38 @@ const COMMENT_PART = 7;
...
  */
-export type TemplateResult<T extends ResultType = ResultType> = {
+export type UncompiledTemplateResult<T extends ResultType = ResultType> = {
   // This property needs to remain unminified.
   ['_$litType$']: T;
   strings: TemplateStringsArray;
   values: unknown[];
 };
 
+export type MaybeCompiledTemplateResult<T extends ResultType = ResultType> =
```
Comment (AndrewJakubowicz) [inline packages/lit-html/src/lit-html.ts:476]: Nit: This could use a JSDoc to clarify why this is used. Probably want to call out that this should ideally be the type of `html` or `svg` when you're trying to target both uncompiled and compiled environments.

File: packages/lit-starter-ts/src/test/my-element_test.ts [hunk 3]
```diff
@@ -16,7 +17,9 @@ suite('my-element', () => {
   });
 
   test('renders with default values', async () => {
-    const el = await fixture(html`<my-element></my-element>`);
+    const el = await fixture(
+      html`<my-element></my-element>` as UncompiledTemplateResult
+    );
```
Comment (AndrewJakubowicz) [inline packages/lit-starter-ts/src/test/my-element_test.ts:22]: I'm not clear why the changes in `my-element_test.ts` are needed? It seems outside the scope of `lit-html`?

File: packages/lit-html/src/lit-html.ts (hunk 2 above)
Comment (rictic) [reply packages/lit-html/src/lit-html.ts:485]: Good point! Done

File: packages/lit-html/src/lit-html.ts (hunk 1 above)
Comment (rictic) [reply packages/lit-html/src/lit-html.ts]: Oh nice!

File: packages/lit-starter-ts/src/test/my-element_test.ts (hunk 3 above)
Comment (rictic) [reply packages/lit-starter-ts/src/test/my-element_test.ts:22]: As part of this change I also fixed all the type errors that would arise if we changed TemplateResult to be MaybeCompiledTemplateResult

**Review by AndrewJakubowicz — APPROVED** [review]
Nice!
